core/src/main/java/com/exadel/aem/backpack/core/services/impl/PackageServiceImpl.java

```diff
@@ -1047,4 +1072,55 @@ private boolean isFolder(String resourceType) {
         return packageInfos;
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public JcrPackageWrapper uploadPackage(final Session session,
+                                           final byte[] fileUploadBytesArray,
+                                           final boolean forceUpdate) {
+        JcrPackageWrapper jcrPackageWrapper = new JcrPackageWrapper();
+        File fileUpload = null;
+
+        if (session != null && fileUploadBytesArray != null) {
+            JcrPackageManager packageManager = getPackageManager(session);
+            try {
+                fileUpload = getFile(fileUploadBytesArray);
+                String nameHint = "uploaded_package_name";
+                final boolean isTempFile = true;
+                final boolean strict = true;
+
+                JcrPackage upload = packageManager.upload(fileUpload, isTempFile, forceUpdate, nameHint, strict);
+                jcrPackageWrapper.setJcrPackage(upload);
+
+                return jcrPackageWrapper;
+            } catch (Exception e) {
+                LOGGER.error("Cannot upload package: {}", e.getMessage(), e);
+                jcrPackageWrapper.setMessage(e.getMessage());
+                jcrPackageWrapper.setStatusCode(SC_CONFLICT);
+            } finally {
+                if (fileUpload != null) {
+                    fileUpload.delete();
+                }
+            }
+        } else {
+            jcrPackageWrapper.setMessage("An incorrect value of parameter(s)");
+            jcrPackageWrapper.setStatusCode(SC_CONFLICT);
+        }
+
+        return jcrPackageWrapper;
+    }
+
+    private File getFile(final byte[] fileUpload) throws IOException {
+        File tmpFile = allocateTmpFile();
```

Comment:
In case of exception, you will have an open stream. You can add try with resources here.

Reply:
done